### PR TITLE
feat(schema): add scoped user secret names

### LINKED ISSUE #978: Allow restricting which user secrets are exposed to workflow jobs and interactive sessions
After #805, user secrets can be propagated to interactive sessions. Today, secrets are generally exposed broadly to workflow job pods, Dask workers, and interactive sessions. This is convenient, but it also means pods receive secrets they do not actually need.

## Problem

There is currently no least-privilege way for a workflow author or an interactive-session caller to restrict which user secrets are exposed to a given pod/session.

We should keep the current default behaviour for backward compatibility, but provide an explicit opt-in mechanism to reduce the exposed secret set when desired.

## Proposed solution

Introduce an optional `secret_names` allowlist.

Behaviour:

- If `secret_names` is omitted: keep current behaviour and expose all user secrets.
- If `secret_names` is an empty list: expose no user secrets.
- If `secret_names` is provided: expose only the named secrets.
- If a requested secret name does not exist: reject the request with a clear validation error.

Implementation approach:

- Filter by secret name once in `UserSecrets`, then derive env refs and mounted file items from the filtered subset.
- Preserve the current behaviour for Kerberos/VOMS/Rucio-enabled paths for now, i.e. do not scope secrets there yet.

## User-facing scope

Interactive sessions:

- `reana-client open ... --secret-name NAME` (repeatable)
- `reana-client open ... --no-secrets`
- `reana-client-go open ... --secret-name NAME` (repeatable)
- `reana-client-go open ... --no-secrets`

Workflow authoring:

- Serial: step-level `secret_names: ["alpha", "beta"]` in `reana.yaml`
- Dask: `workflow.resources.dask.secret_names: ["alpha", "beta"]` in `reana.yaml`
- Snakemake: rule-level `resources: secret_names="alpha,beta"`
- Yadage: step `environment.resources` entry such as `- secret_names: ["alpha", "beta"]`

## Out of scope for this issue

- CWL support
- Secret scoping for Kerberos/VOMS/Rucio-enabled jobs

These can be addressed in follow-up work once the base scoping model is agreed.

## Acceptance criteria

- Existing workflows and interactive-session flows remain unchanged when `secret_names` is not provided.
- Interactive sessions can request either an explicit secret allowlist or no secrets at all.
- Serial, Snakemake, Yadage, and Dask authors can specify the secret subset for their jobs/workers.
- Empty allowlists result in no secrets being exposed.
- Unknown secret names produce a clear user-facing error.
- Kerberos/VOMS/Rucio-enabled paths preserve current behaviour.

## Affected components

- `reana-commons`
- `reana-job-controller`
- `reana-workflow-controller`
- `reana-client`
- `reana-client-go`
- `reana-workflow-engine-serial`
- `reana-workflow-engine-snakemake`
- `reana-workflow-engine-yadage`
- `reana-server`

### PR DESCRIPTION
**Pull request by michaelbuchar**
Propagate optional secret allowlists through the shared schemas and API client, and add workflow-aware secret discovery helpers that preserve omitted-vs-empty semantics, step overrides, CWL hints, and feature credential carve-outs.

Closes reanahub/reana#978

### CODE REVIEW

File: reana_commons/k8s/secrets.py [hunk 1]
```diff
@@ -26,6 +27,336 @@
 from reana_commons.utils import build_unique_component_name
 
 log = logging.getLogger(__name__)
+_CERN_BACKENDS_REQUIRING_KERBEROS = {"htcondorcern", "slurmcern"}
+_MISSING = object()
+
+
+def get_workflow_secret_names(
```
Comment (CameronMcClymont) [inline reana_commons/k8s/secrets.py:38]: For file-type secrets, the pod spec attaches a volume pointing at the whole secret bag, plus an optional `items` list meaning which secrets in the bag should be projected as files. Before this PR set, the whole secrets bag would be attached. This PR set allows scoping using `secret_names:` before building the `items` list mentioned above.

When your `secret_names` list only contains env-type secrets (no file-type secrets), this `items` list comes out empty. The problem is that Kubernetes treats an empty `items` list to mean "include the whole secrets bag" rather than "don't include anything from the secrets bag". In this PR set, the latter is the convention so it's easy to overlook.

The tests miss it because every test that inspects the mounted files happens to put a file secret in the allowlist, which makes `items` non-empty and hides the behaviour.

The fix is small: decide whether to attach the secrets volume by "are there file secrets in scope" rather than "are there any secrets in scope".

File: reana_commons/k8s/secrets.py [hunk 2]
```diff
@@ -26,6 +27,336 @@
...
+    if isinstance(scope.get("process"), dict) and isinstance(
+        scope.get("environment"), dict
+    ):
+        for resource in scope.get("environment", {}).get("resources", []) or []:
+            if (
+                feature_name == "kerberos"
+                and resource.get("compute_backend") in _CERN_BACKENDS_REQUIRING_KERBEROS
+            ):
+                return True
+            if isinstance(resource, dict) and resource.get(feature_name):
```
Comment (CameronMcClymont) [inline reana_commons/k8s/secrets.py]: This `isinstance` check should be the first `if` statement in this loop to ensure that `resource.get("compute_backend")` above doesn't fail with `AttributeError: 'str' object has no attribute 'get'`. Could also add a regression test for this using e.g. `resources: ["GRIDProxy"]`.

File: reana_commons/openapi_specifications/reana_job_controller.json [hunk 3]
```diff
@@ -1,7 +1,6 @@
 {
   "definitions": {
     "Job": {
-      "additionalProperties": false,
```
Comment (CameronMcClymont) [inline reana_commons/openapi_specifications/reana_job_controller.json:4]: Before this PR, this file and `reana-job-controller/docs/openapi.json` differed only by a version string. After this PR they diverge so the reana-commons copy drops "additionalProperties": false from Job and JobRequest, drops "additionalProperties": {} from env_vars, and gains "format": "int32" on four integer fields. None of that appears in the job-controller PR's own spec.

File: reana_commons/k8s/secrets.py [hunk 4]
```diff
@@ -26,6 +27,336 @@
...
+    scoped_secret_names: Optional[List[str]],
+    workflow_resources: Optional[Dict[str, Any]] = None,
+) -> Optional[List[str]]:
+    """Resolve a scope-local allowlist against the workflow-global default."""
+    if scoped_secret_names is not None:
+        return scoped_secret_names
+    return get_workflow_secret_names(workflow_resources)
+
+
+def _normalize_secret_names(secret_names: Any) -> Optional[List[str]]:
```
Comment (CameronMcClymont) [inline reana_commons/k8s/secrets.py:58]: `_normalize_secret_names()` returns `[]` for any malformed input and non-string members of a list are dropped without comment. The JSON schema catches this for serial steps and `workflow.resources`, but yadage `environment.resources` and Snakemake rule resources are not schema-validated. Instead, raise a validation error in these cases instead of coercing.

**Review by CameronMcClymont — CHANGES_REQUESTED** [review]

File: reana_commons/k8s/secrets.py (hunk 1 above)
Comment (michaelbuchar) [reply reana_commons/k8s/secrets.py:38]: Fixed

File: reana_commons/k8s/secrets.py (hunk 2 above)
Comment (michaelbuchar) [reply reana_commons/k8s/secrets.py]: Fixed

File: reana_commons/openapi_specifications/reana_job_controller.json (hunk 3 above)
Comment (michaelbuchar) [reply reana_commons/openapi_specifications/reana_job_controller.json:4]: Fixed

File: reana_commons/k8s/secrets.py (hunk 4 above)
Comment (michaelbuchar) [reply reana_commons/k8s/secrets.py:58]: Fixed

File: reana_commons/k8s/secrets.py [hunk 5]
```diff
@@ -21,11 +22,348 @@
...
+    workflow_resources: Optional[Dict[str, Any]] = None,
+) -> Optional[List[str]]:
+    """Return the workflow-global secret allowlist if configured."""
+    workflow_resources = workflow_resources or {}
+    if "secret_names" in workflow_resources:
+        return _normalize_secret_names(workflow_resources["secret_names"])
+    return None
+
+
+def resolve_secret_names(
```
Comment (tiborsimko) [inline reana_commons/k8s/secrets.py:48]: **🤖 PR546-01: Engine-facing secret-scoping helpers require the Kubernetes extra (HIGH)**

`resolve_secret_names()` and the workflow-scope discovery helpers are pure data-processing functions, but they now live in `reana_commons.k8s.secrets`. Importing that module immediately imports `kubernetes`, which is only provided by the optional `reana-commons[kubernetes]` extra. The Serial, CWL, and Yadage engines import these helpers while depending on `reana-commons`, `reana-commons[cwl]`, or `reana-commons[yadage]`; none installs the Kubernetes client. Consequently these three engines fail during module import with `ModuleNotFoundError: No module named 'kubernetes'`. Publishing the amended commons package will not make those downstream jobs green. Snakemake happens to be unaffected because its Kubernetes executor dependency supplies the package.

Suggested fix — move the dependency-free scoping functions (`resolve_secret_names()`, normalization, and workflow-scope discovery) into a Kubernetes-independent commons module, update the engines to import from there, and optionally re-export them from `k8s.secrets` for compatibility. Please add an import-level test using only each engine's declared extras so this dependency boundary remains enforced; adding the Kubernetes client to engines that do not use the Kubernetes API would unnecessarily broaden their dependency surface.

File: reana_commons/snakemake.py [hunk 6]
```diff
@@ -71,6 +71,13 @@ def snakemake_load(workflow_file: str, **kwargs: Any):
         return snakemake_load_v7(workflow_file, **kwargs)
 
 
+def parse_secret_names_resource(value: Any) -> Optional[List[str]]:
+    """Parse Snakemake's scalar ``secret_names`` resource into a list."""
+    if value is None:
+        return None
+    return [secret.strip() for secret in str(value).split(",") if secret.strip()]
```
Comment (tiborsimko) [inline reana_commons/snakemake.py:78]: **🤖 PR546-02: Snakemake silently coerces malformed secret declarations (LOW)**

`parse_secret_names_resource()` converts every non-`None` value to `str` before splitting it. This means malformed declarations such as `secret_names=123` or `secret_names=True` are silently accepted as secret names `"123"` and `"True"`. Snakemake rule resources are not covered by the `reana.yaml` JSON schema, so this parser is the validation boundary. The shared `_normalize_secret_names()` correctly rejects the equivalent malformed values, making the user-visible contract depend on which workflow engine is used.

Suggested fix — require a string value here and raise `REANAValidationError` for other types before splitting. Add regression coverage for integer, boolean, and valid comma-separated inputs.

File: reana_commons/k8s/secrets.py [hunk 7]
```diff
@@ -21,11 +22,348 @@
...
+                    continue
+                run_definition = step.get("run")
+                tool = run_definition if isinstance(run_definition, dict) else None
+                yield _build_effective_cwl_step_scope(step, tool)
+            for key, value in workflow_specification.items():
+                if key != "steps":
+                    yield from _iter_executable_scopes(value)
+            return
+
+        if isinstance(workflow_specification.get("commands"), list):
```
Comment (tiborsimko) [inline reana_commons/k8s/secrets.py:235]: **🤖 PR546-04: Whole-tree scope detection can silently expose all secrets to the sidecar (MEDIUM)**

`_iter_executable_scopes()` recursively walks the entire stored workflow specification and treats any mapping with a list-valued `commands` key as an executable Serial/Snakemake scope. Valid workflow data can contain the same shape outside an executable step—for example, a Snakemake input or parameter named `commands` whose value is a list. Such data is then counted as an unscoped executable scope. If all real steps have local allowlists and there is no workflow-global default, `get_declared_workflow_secret_names()` changes from their union to `None`, which means exposing every user secret to the job-controller sidecar. With a global default, it unnecessarily widens the sidecar union instead. Job pods remain locally scoped, but the sidecar's least-privilege boundary is silently defeated. The unrestricted recursive walk can also exhaust Python's recursion limit on deeply nested user-controlled data.

This is reproducible at the current head: for a Serial specification whose executable steps all declare `secret_names: ["S1"]`, adding unrelated nested data `{"commands": ["echo hi"]}` changes the result from `["S1"]` to `None` when no workflow-global default is set. With a global `secret_names: ["G1"]` default, the same nested data widens the result from `["S1"]` to `["G1", "S1"]`.

Suggested fix — make scope traversal workflow-type-aware and inspect only the known executable locations for Serial, Snakemake, Yadage, and CWL rather than applying shape heuristics to the whole document. Prefer an iterative or explicitly bounded traversal where recursion is still needed. Add regression tests with valid nested data containing `commands: [...]` and with deeply nested non-executable data.

File: reana_commons/k8s/secrets.py (hunk 1 above)
Comment (tiborsimko) [reply reana_commons/k8s/secrets.py:38]: **🤖 Follow-up (2026-08-23): verified at current head**

The amended code adds `has_file_secrets()` and guards the production secret-volume attachments with it. Env-only and explicitly empty scopes therefore no longer attach a Secret volume with `items: []`. The added tests also exercise env-only scoping. This addresses the behavior raised in this thread; I consider this discussion resolved.

The conditional attachment also exposes a separate mismatch in the Rucio init-container callers: reana-job-controller and reana-workflow-controller still attach the corresponding mount unconditionally. I have raised those independently as PR525-01 and PR692-01; they do not reopen this thread's original empty-projection concern.

File: reana_commons/k8s/secrets.py (hunk 2 above)
Comment (tiborsimko) [reply reana_commons/k8s/secrets.py]: **🤖 Follow-up (2026-08-23): verified at current head**

The resource-type guard now runs before any `.get()` access, and the regression coverage includes a non-mapping Yadage resource entry. This addresses the `AttributeError` path raised here; I consider this discussion resolved.

File: reana_commons/openapi_specifications/reana_job_controller.json (hunk 3 above)
Comment (tiborsimko) [reply reana_commons/openapi_specifications/reana_job_controller.json:4]: **🤖 Follow-up (2026-08-23): verified at current head**

The reana-commons and reana-job-controller OpenAPI copies are synchronized again; the corresponding documents differ only where their component version is expected to differ. This addresses the accidental schema drift raised here; I consider this discussion resolved.

File: reana_commons/k8s/secrets.py (hunk 4 above)
Comment (tiborsimko) [reply reana_commons/k8s/secrets.py:58]: **🤖 Follow-up (2026-08-23): partially addressed**

`_normalize_secret_names()` now raises `REANAValidationError` for non-string values and mixed lists, which fixes the silent coercion in the shared normalizer. Two user-facing paths remain incomplete: Snakemake still applies `str(value)` before parsing its rule resource, so integers and booleans are accepted as names, and a validation error raised while starting a workflow is not mapped to HTTP 400 by the workflow-controller status endpoint and therefore becomes HTTP 500. PR546-02 and PR692-03 track these remaining paths. I would keep this thread open until malformed engine-specific declarations are consistently rejected as client errors.

File: reana_commons/k8s/secrets.py [hunk 8]
```diff
@@ -21,11 +22,348 @@
...
+        else:
+            effective_scope[field_name] = value
+
+    # The effective scope already carries the resolved REANA fields above.
+    effective_scope["hints"] = []
+    effective_scope["requirements"] = []
+    return effective_scope
+
+
+def _iter_packed_cwl_step_scopes(workflow_specification: Dict[str, Any]):
```
Comment (tiborsimko) [inline reana_commons/k8s/secrets.py:161]: **🤖 PR546-07: CWL job and sidecar secret scopes use incompatible hint resolution (MEDIUM)**

cwltool inherits hints from an enclosing `Workflow` into its steps and tools. `ReanaPipelineJob._get_hint()` then scans that inherited list from the beginning, so the engine currently resolves duplicate REANA fields in workflow → step → tool order. By contrast, `_build_effective_cwl_step_scope()` models only step → tool and never considers the enclosing Workflow node. cwltool's native `get_requirement()` applies a third order, tool → step → workflow, by scanning the inherited list in reverse.

The missing workflow layer already defeats least privilege without any conflicting declarations. A workflow-level REANA hint containing `secret_names: ["S1"]` reaches each `ReanaPipelineJob`, so the submitted jobs use `S1`; commons sees every packed step as unscoped and `get_declared_workflow_secret_names()` returns `None`, exposing every user secret to the job-controller sidecar. With declarations at multiple levels, the engine and sidecar can also choose different scopes, and the engine's forward scan prevents a step from narrowing a workflow-level value. The two consumers additionally disagree on class matching: the engine accepts any hint containing the requested field, while commons accepts only an absent class or `class: reana`.

This is reproducible with the packed `$graph` representation stored by REANA. With `secret_names: ["S1"]` only on the Workflow node, sidecar discovery returns `None`; moving the same hint to the step returns `["S1"]`. With `secret_names: ["WORKFLOW_LEVEL"]` on the Workflow and `["STEP_LEVEL"]` on the step, the inherited engine hint list contains both values in that order and `_get_hint("secret_names")` returns the workflow value.

The forward-scan precedence is pre-existing and applies to every REANA CWL hint read through `_get_hint()`, including `compute_backend`, `kerberos`, `voms_proxy`, `rucio`, and `kubernetes_uid`. Changing it for secret scoping can therefore alter other scheduling and credential behavior, so those fields need coordinated regression coverage as part of the resolver fix.

Suggested fix — define the intended workflow/step/tool precedence for REANA CWL fields and implement it through one shared resolver used by both workflow-scope discovery in reana-commons and runtime job construction in reana-workflow-engine-cwl. Preserve the local-over-global scoping model when a shared tool is invoked from steps with different allowlists. Add end-to-end packed-CWL tests for a workflow-only hint, conflicting workflow/step/tool declarations, nested workflows, an unrelated foreign hint class, and the existing non-secret REANA fields listed above.

File: reana_commons/k8s/secrets.py (hunk 5 above)
Comment (tiborsimko) [reply reana_commons/k8s/secrets.py:48]: **🤖 Follow-up:** I rechecked this against the current heads. It remains a deterministic blocker rather than a release-ordering failure: Serial, CWL, and Yadage do not install the Kubernetes client through their declared dependencies, so their engine imports fail before execution. Publishing the amended commons package alone will not resolve these failures. Moving the pure scoping helpers to a Kubernetes-independent module would unblock all three downstream PRs.
